multicluster/controllers/multicluster/node_controller.go

```diff
@@ -73,67 +78,215 @@ func NewNodeReconciler(
+// during controller startup.
+func (r *NodeReconciler) initializeActiveGateway() error {
+	ctx := context.Background()
+	nodeList := &corev1.NodeList{}
+	if err := r.Client.List(ctx, nodeList, &client.ListOptions{}); err != nil {
+		return err
+	}
+	// Empty the cache in case of initializeActiveGateway failed and rerun.
+	r.gatewayCandidates = make(map[string]string)
+	for _, n := range nodeList.Items {
+		if _, isGW := n.Annotations[common.GatewayAnnotation]; isGW {
+			r.gatewayCandidates[n.Name] = n.Name
+		}
+	}
+
+	gwList := &mcsv1alpha1.GatewayList{}
+	if err := r.Client.List(ctx, gwList, &client.ListOptions{}); err != nil {
+		return err
+	}
+	// Gateway webhook guarantees that there is at most one Gateway in the member cluster.
+	size := len(gwList.Items)
+	if size == 1 {
+		existingGWName := gwList.Items[0].Name
+		// Check existing Gateway Node's health
+		var nodeNotFound bool
+		node := &corev1.Node{}
+		if err := r.Client.Get(ctx, types.NamespacedName{Name: existingGWName}, node); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			nodeNotFound = true
+		}
+		_, _, err := r.getGatawayNodeIP(node)
+		// Delete the existing Gateway if it's no longer valid.
+		if !isReadyGateway(node) || err != nil || nodeNotFound {
+			if err != nil {
+				klog.ErrorS(err, "There is no valid Gateway IP for Node", "node", node.Name)
+			}
+			return r.Client.Delete(ctx, &mcsv1alpha1.Gateway{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: r.namespace,
+					Name:      existingGWName,
+				}})
+		}
+		r.activeGateway = existingGWName
+	}
+	// Other cases will be handled by reconciler after initialization.
+	return nil
+}
+
+func (r *NodeReconciler) updateActiveGateway(ctx context.Context, newGateway *mcsv1alpha1.Gateway) error {
+	existingGW := &mcsv1alpha1.Gateway{}
+	// TODO: cache might be stale. Need to revisit here and other reconcilers to
+	// check if we can improve this with 'Owns' or other methods.
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: newGateway.Name, Namespace: r.namespace}, existingGW); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+	existingGW.GatewayIP = newGateway.GatewayIP
+	existingGW.InternalIP = newGateway.InternalIP
+	if err := r.Client.Update(ctx, existingGW, &client.UpdateOptions{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// newActiveGateway will delete existing Gateway CR and create a new Gateway
+// from the pool of Gateway candidates.
+func (r *NodeReconciler) recreateActiveGateway(ctx context.Context, gateway *mcsv1alpha1.Gateway) error {
+	err := r.Client.Delete(ctx, gateway, &client.DeleteOptions{})
+	if err == nil || apierrors.IsNotFound(err) {
+		r.activeGateway = ""
+		delete(r.gatewayCandidates, gateway.Name)
+		// Check remaining Gateway's candidates and create a new Gateway.
+		return r.newActiveGatewayFromCandidates()
+	}
+	return err
+}
+
+func (r *NodeReconciler) newActiveGatewayFromCandidates() error {
```

Comment:
Could you add comments that the func will return no error if no good Gateway candidate?

Reply:
Done